### PR TITLE
BC-2923 - elements anywhere

### PR DESCRIPTION
**Pull request by christian-darsow**
# Short Description

Create new (text) element after each existing (text) element and not only at the bottom of the elements list

Refactoring:
- move elements' list from page to own component
- rename CardElement to CardElementWrapper

## Links to Ticket and related Pull-Requests
https://ticketsystem.dbildungscloud.de/browse/BC-2923

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/components/card-elements/AddCardElement.vue [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	<v-card flat class="dashed mb-4">
+		<v-col class="text-center">
+			<v-btn
+				icon
+				outlined
+				color="secondary"
+				class="add-element-btn"
+				data-testid="add-element-btn"
+				@click="addElement"
+			>
```
Comment (odalys-dataport) [inline src/components/card-elements/AddCardElement.vue:11]: the button is missing an aria-label

File: src/components/card-elements/AddCardElement.vue [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	</v-card>
+</template>
+
+<script lang="ts">
+import { defineComponent } from "vue";
+import { mdiPlus } from "@mdi/js";
+
+export default defineComponent({
+	name: "AddCardElement",
+	components: {},
```
Comment (odalys-dataport) [inline src/components/card-elements/AddCardElement.vue:24]: pls remove

File: src/components/card-elements/CardElementList.vue [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import VueI18n from "vue-i18n";
+import CardElementWrapper from "@/components/card-elements/CardElementWrapper.vue";
+import AddCardElement from "@/components/card-elements/AddCardElement.vue";
+import { CardElementComponentEnum } from "@/store/types/card-element";
+import { CardElementResponseCardElementTypeEnum } from "@/serverApi/v3";
+import { useDrag } from "@/composables/drag";
+import draggable from "vuedraggable";
+
+export default defineComponent({
+	name: "CardElement",
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementList.vue:43]: name should be CardElementList

File: src/components/card-elements/CardElementList.vue [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+				model: "",
+				props: {
+					component: CardElementComponentEnum.RichText,
+					placeholder: "common.labels.description",
+					editable: true,
+				},
+			});
+		};
+
+		const addElement = () => {
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementList.vue:77]: Could we remove this method and give addElementAfter a default parameter?

File: src/components/card-elements/CardElementWrapper.vue [hunk 5]
```diff
@@ -0,0 +1,171 @@
+<template>
+	<v-card flat class="card mb-7" :class="{ active: isActive }">
+		<div class="card-actions">
+			<v-btn
+				v-show="actionable"
+				icon
+				outlined
+				color="secondary"
+				class="drag-element-btn handle"
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:9]: missing aria-label

File: src/components/card-elements/CardElementWrapper.vue [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+				data-testid="drag-element-btn"
+			>
+				<v-icon>{{ mdiDrag }}</v-icon>
+			</v-btn>
+			<v-btn
+				v-show="actionable && isActive"
+				icon
+				outlined
+				color="secondary"
+				class="delete-element-btn"
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:19]: missing aria-label

File: src/components/card-elements/CardElementWrapper.vue [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+		<component
+			:is="component"
+			v-model="model"
+			:editable="editable"
+			:disabled="disabled"
+			:placeholder="`${$t(placeholder)}`"
+			@input="handleInput"
+			@focus="handleFocus"
+			@blur="handleBlur"
+		></component>
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:35]: pls remove empty closing tag

File: src/components/card-elements/CardElementWrapper.vue [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		<v-btn
+			v-show="actionable && isActive"
+			icon
+			outlined
+			color="secondary"
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:40]: missing aria-label

File: src/components/card-elements/CardElementWrapper.vue [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+<script lang="ts">
+import { defineComponent, ref, watch, inject, computed } from "vue";
+import { mdiTrashCanOutline, mdiDrag, mdiPlus } from "@mdi/js";
+import { CardElementComponentEnum } from "@/store/types/card-element";
+import TitleCardElement from "@/components/card-elements/TitleCardElement.vue";
+import RichTextCardElement from "@/components/card-elements/RichTextCardElement.vue";
+import VueI18n from "vue-i18n";
+
+export default defineComponent({
+	name: "CardElement",
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:59]: name should be CardElementWrapper

File: src/components/card-elements/CardElementList.vue [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	},
+	emits: ["input"],
+	props: {
+		value: {
+			type: Array,
+			default: () => [],
+		},
+	},
+	setup(props, { emit }) {
+		const i18n: VueI18n | undefined = inject<VueI18n>("i18n");
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementList.vue:57]: i18n is injected but I don't see any usage of it

File: src/components/card-elements/CardElementWrapper.vue [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	components: {
+		TitleCardElement,
+		RichTextCardElement,
+	},
+	emits: ["input", "delete-element", "add-element"],
+	props: {
+		component: {
+			type: String,
+			default: CardElementComponentEnum.Title,
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:68]: Is this a good default, if Title is only supposed to exist once?

File: src/components/card-elements/CardElementWrapper.vue [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+.drag-element-btn {
+	position: relative;
+	top: -#{$btn-radius}px;
+}
+
+.delete-element-btn {
+	position: relative;
+	/* stylelint-disable-next-line sh-waqar/declaration-use-variable */
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:160]: after the nuxt removal we don't have stylelint anymore, so these should be unneccesary

File: src/components/card-elements/CardElementWrapper.vue [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+		},
+		disabled: {
+			type: Boolean,
+		},
+		editable: {
+			type: Boolean,
+		},
+	},
+	setup(props, { emit }) {
+		const i18n: VueI18n | undefined = inject<VueI18n>("i18n");
```
Comment (odalys-dataport) [inline src/components/card-elements/CardElementWrapper.vue:86]: i18n is injected but I don't see any usage of it

File: src/pages/TaskCard.page.vue [hunk 14]
```diff
@@ -99,7 +78,7 @@ export default defineComponent({
 
 		onMounted(async () => {
 			const taskCardId =
-				route.name === "task-card-edit" ? route.params.id : undefined;
+				route.name === "task-edit" ? route.params.id : undefined;
```
Comment (odalys-dataport) [inline src/pages/TaskCard.page.vue:81]: shouldn't we rather rename the route names?

**Review by odalys-dataport — CHANGES_REQUESTED** [review]

File: src/components/card-elements/CardElementList.vue (hunk 10 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementList.vue:57]: Fixed that and used for translation of the description

File: src/components/card-elements/CardElementList.vue (hunk 4 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementList.vue:77]: Nice :) changed and removed dumb addElement

File: src/components/card-elements/CardElementWrapper.vue (hunk 7 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementWrapper.vue:35]: done

File: src/components/card-elements/AddCardElement.vue (hunk 2 above)
Comment (christian-darsow) [reply src/components/card-elements/AddCardElement.vue:24]: done

File: src/components/card-elements/CardElementList.vue (hunk 3 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementList.vue:43]: fixed

File: src/components/card-elements/CardElementWrapper.vue (hunk 9 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementWrapper.vue:59]: omg. One more time. fixed

File: src/components/card-elements/CardElementWrapper.vue (hunk 11 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementWrapper.vue:68]: Very good point indeed! changed this to our text component for now. 
Maybe we should re-open our discussion in generel and talk of title is a card element at all

File: src/components/card-elements/CardElementWrapper.vue (hunk 13 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementWrapper.vue:86]: yeap, removed this

File: src/components/card-elements/CardElementWrapper.vue (hunk 12 above)
Comment (christian-darsow) [reply src/components/card-elements/CardElementWrapper.vue:160]: Thanks!

File: src/pages/TaskCard.page.vue (hunk 14 above)
Comment (christian-darsow) [reply src/pages/TaskCard.page.vue:81]: Agree, changed the routes names back to what they were before

File: src/components/card-elements/CardElementWrapper.vue (hunk 11 above)
Comment (odalys-dataport) [reply src/components/card-elements/CardElementWrapper.vue]: I agree

**Review by odalys-dataport — APPROVED** [review]

**Pull request merged**